src/app/qgshandlebadlayers.cpp

```diff
@@ -509,3 +495,164 @@ int QgsHandleBadLayers::layerCount()
 {
   return mLayerList->rowCount();
 }
+
+QString QgsHandleBadLayers::findFile( const QString &fileName, const QString &basePath, int maxDepth, int driveMargin )
+{
+  int depth = 0;
+  QDir folder = QDir( QDir( basePath ).absolutePath() );
+  QString existingBase;
+  // find the nearest existing folder
+  while ( !folder.exists() && folder.absolutePath().count( '/' ) > driveMargin )
+  {
+    existingBase = folder.path();
+    if ( !folder.cdUp() )
+      folder = QFileInfo( existingBase ).absoluteDir(); // using fileinfo to move up one level
+    depth += 1;
+  }
```

Comment:
I'm a bit shaky on the logic here -- can you confirm I have it correct?
- at this stage we've trimmed subfolders off the original folder path until we've eventually hit a folder which actually still exists (and I think we should be returning out early here if that didn't happen)
- we then keep traversing up through the folders, at each stage searching through all subdirectories looking for the file?

If so, I can understand the intention. But my concern is that we're duplicating searches, since every time we move up a folder we're also re-checking the subfolder (and it's contents) which we just came from.

Reply:
Yes this is the case and I agree that there is some wasted potential, it might be possible to skip the folder that was already 'scanned' but I would need to add another loop for that if you don't mind.